### ISSUE TITLE
Make --remote_local_fallback honor --spawn_strategy

### ISSUE THREAD

**Issue by ulrfa** [bazelbuild/bazel#15519]
### Description of the feature request:

In case of remote execution issues, I expect `--remote_local_fallback` with `--spawn_strategy=remote,worker,linux-sandbox` to fallback to `worker` if available, and otherwise to `linux-sandbox`.

But instead I see fallback based on `--remote_local_fallback_strategy`, which is supposed to be a **no-op** according to [documentation](https://bazel.google.cn/reference/command-line-reference?hl=en#flag--remote_local_fallback_strategy)

I see two alternatives, preferably:

- make `--remote_local_fallback` honor `--spawn_strategy`

or otherwise:

- Update documentation for `--remote_local_fallback_strategy` about it being a no-op (which is incorrect) and remove the deprecation of it.

### What underlying problem are you trying to solve with this feature?

I would like reliable builds with linux-sandbox, when remote execution cluster is temporarily unavailable.

I could use `--remote_local_fallback_strategy` despite it claiming to be a no-op, but that results in deprecation warnings.

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

I use a custom build with local patches on top of 5.0.0-pre.20210929.1, but it seems this code has not changed on main branch.

### Have you found anything relevant by searching the web?

Based on description/comments from @agoulti and @philwo in https://github.com/bazelbuild/bazel/issues/5400, it seems the intention was that `--remote_local_fallback_strategy` should not be needed any more.

Then @ishikhman wrote a https://github.com/bazelbuild/bazel/issues/7480#issuecomment-528352938 that it is still needed for this use case, but ticket was closed without addressing that.

**Comment by larsrc-google** [comment]
I've been looking at the fallback situation for a bit internally, where there are some extra legacy flags. `--remote_local_fallback` has the very reasonable use case of handling certain remote failures. Getting rid of `--remote_local_fallback_strategy` would be good. As for what should be the fallback strategy, I think it should just go on to the next strategy in its list of possible strategies - just as if remote execution had not been allowed for the action.

Example (with made-up error messages):
```
bazel build --spawn_strategy=remote,worker,local --strategy=Foo=remote,local //foo:bar
...
INFO: Foo action failed remotely, falling back to `local` strategy.
INFO: Bar action failed remotely, falling back to `worker` strategy.
...
```

**Comment by ulrfa** [comment]
> I've been looking at the fallback situation for a bit internally, where there are some extra legacy flags. `--remote_local_fallback` has the very reasonable use case of handling certain remote failures. Getting rid of `--remote_local_fallback_strategy` would be good. As for what should be the fallback strategy, I think it should just go on to the next strategy in its list of possible strategies - just as if remote execution had not been allowed for the action.
> 
> Example (with made-up error messages):
> 
> ```
> bazel build --spawn_strategy=remote,worker,local --strategy=Foo=remote,local //foo:bar
> ...
> INFO: Foo action failed remotely, falling back to `local` strategy.
> INFO: Bar action failed remotely, falling back to `worker` strategy.
> ...
> ```

Thanks for working on this @larsrc-google! I think your solution is good!

My only concern is the number of messages printed in terminal. Are you planning reporting each individual action? I'm afraid that is too verbose when remote system is down and all actions fallback. Or is it just me misunderstanding the example?

**Comment by larsrc-google** [comment]
No, those were just for illustration. In practice, they should go in the log only.

**Comment by liyuyun-lyy** [comment]
Agree. The description of the [document](https://github.com/bazelbuild/bazel/issues/7480) is quite confusing. Actually, we don't need to specify the spawn_strategy because the  default strategy is already `--spawn_strategy=remote,worker,linux-sandbox`.
However, it has nothing to do with `--remote_local_fallback`, and it cannot replace `--remote_local_fallback_strategy`. I'm surprised to find out that default local_fallback_strategy is local, because most of the time, the default strategy for Bazel is sandboxed.

**Comment by liyuyun-lyy** [comment]
> remote_local_fallback_strategy

I'll give three  examples.
1. we only set `--spawn_strategy=remote,worker,linux-sandbox`.

**Comment by larsrc-google** [comment]
Ah, so you're saying that the current implementation doesn't help. But what if `--remote_local_fallback` was changed to mean "try the next strategy in the regular list of strategies"?

**Comment by liyuyun-lyy** [comment]
> Ah, so you're saying that the current implementation doesn't help. But what if `--remote_local_fallback` was changed to mean "try the next strategy in the regular list of strategies"?

Yes, I think that will be better.

**Comment by larsrc-google** [comment]
Hmmm... that would take some rework, Right now `exec() (src/main/java/com/google/devtools/build/lib/exec/SpawnStrategyResolver.java:43) just picks the first possible strategy and execs that, so we don't really know where we are in the list of possible strategies. The `--remote_local_fallback` handling is entirely separate (src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnRunner.java:485).

**Comment by larsrc-google** [comment]
One thing we're missing is a way to signal "let someone else try" in the failure messages. This has also come up in dynamic execution and for workers.

**Comment by jkeljo** [comment]
Since `--remote_local_fallback_strategy` is not a no-op (setting it to `linux-sandbox` does in fact cause Bazel to fall back to that strategy instead of `local`), what do you all think about updating the docs and removing the `@Deprecated` tag while you're working on figuring out the next steps for spawn strategies as a whole?
